### ISSUE TITLE
Option to disable {{{ }}}

### ISSUE THREAD

**Issue by stevenvachon** [wycats/handlebars.js#1363]
Unescaped output can be dangerous and `{{{ }}}` makes it too easy to accidentally do so. It'd be great if we could optionally disable it, resulting in a thrown template parsing error.

**Comment by stevenvachon** [comment]
@nknapp @wycats

**Comment by nknapp** [comment]
Technically, this is a new feature, so I would like @wycats to agree to this before proceeding this further. But I think this is independent from #1277, since it is not a change in the Handlebars-language itself. 

@stevenvachon 
* Do you need this feature yourself, or do you just think that somebody else might need it?
* Do you already know what needs to be changed in order to achieve this goal? 

A validator could possibly be created easily, using the [compiler-api](https://github.com/wycats/handlebars.js/blob/master/docs/compiler-api.md#ast) with an [AST visitor](https://github.com/wycats/handlebars.js/blob/master/docs/compiler-api.md#ast-visitor).

**Comment by wycats** [comment]
@nknapp I don't have any objection in principle to adding this as an option.

@stevenvachon do you also want to disable `{{` when used with safe strings?

**Comment by nknapp** [comment]
Could be done like this as a seperate module as well https://jsfiddle.net/9D88g/126/

**Comment by stevenvachon** [comment]
@nknapp I need this feature and I know of others in security who avoid Handlebars for this specific reason. In React, for example, they named this feature `dangerouslySetInnerHTML`. I'm not familiar with Handlebars beyond its AST output to know what needs to be changed.

@wycats I'm not sure what you mean.

**Comment by Pisi-Deff** [comment]
@stevenvachon 
I believe he's asking whether returning SafeStrings from helper functions would need to be disabled as well, since they can output unescaped html: http://handlebarsjs.com/#html-escaping

**Comment by stevenvachon** [comment]
@Pisi-Deff or did "safe strings" refer to strings wrapped in an object that contain a flag to indicate that they have been pre-escaped and can therefore be safely used in unsafe contexts?

**Comment by nknapp** [comment]
Like @Pisi-Deff, I also think @wycats meant Handlebars.SafeString-instances returned by helper functions. Should this be forbidden too? 

I would say "no", because this is nothing you do accidentially like writing `{{{` instead of `{{`.

**Comment by stevenvachon** [comment]
I'm not sure what a "safe string" is and how it gets into a helper. Are we referring to a `{{outerHelper}}` that returns `{{{innerHelper}}}` which is interpolated to an unescaped `<tag>`? If so, the `{{{innerHelper}}}` should throw an error if triple-stashes are disabled.

As long as `{{helper}}` *always* escapes its output, it should be fine.

**Comment by Pisi-Deff** [comment]
@stevenvachon 
Please read the last paragraph of: http://handlebarsjs.com/#html-escaping
It describes exactly what a safe string is and how it's used.

**Comment by stevenvachon** [comment]
Ok. `{{` should *always* return a safe (escaped) string. A helper that returns an instance of `Handlebars.SafeString` (which is incorrectly named), should throw an error unless `{{{` is used.

This is a core breaking change to this library, but it is essential for security.

**Comment by jkantr** [comment]
I think there are two separate concerns here and they are both important. Ultimately they are the reason I don't generally recommend handlebars.

1. {{{ }}} can legitimately be done accidentally, it's.. just very similar to {{ }}, especially when your editor auto-closes template syntax for you
2. {{ }} syntactically denotes an escaped string, but semantically allows unescaped strings when a helper returns a certain construct that is encapsulated from the template

As reference, two template engines that "get it right", so to speak, from a security perspective, force by default transparency of the unsafe nature of some interpolation at the template level:

- pug exclusively uses !{foo} syntax for unescaped values
- nunjucks only allows custom filters to return unescaped if piped through a magic filter (eg, {{ foo | safe }})

This always seemed like more of a philosophy direction to me than anything else, but if they are adjustments you are willing to look into, the arguments can certainly be made

refs: https://pugjs.org/language/interpolation.html - https://mozilla.github.io/nunjucks/templating.html#autoescaping

**Comment by nknapp** [comment]
My opinion on this:

## SafeString

The only use-case of `Handlebars.SafeString` is to provide a way for helpers to return strings that need not be escape, because they are already safe (guaranteed by the helper) (see https://jsfiddle.net/9D88g/132/).
I think this use case is valid in principle and should not be removed. 

It is not possible, to throw a ParseError, because this behavior happens in the runtime, when the template is already parsed. If we add an option, to disable `{{{`, it might be possible to disable SafeStrings as well. 
However, it would not throw during compilation, but when the compiled template is executed.

## Unescaped literals

`{{{text}}}` returns the value of `text` without escaping, and I agree that this can happen accidentally.
This can be disabled by the option and a ParseError can be thrown during compilation.

## (No) breaking changes

I don't want to introduce any breaking changes at the moment. I can publish Handlebars 4.1.0 with an option to enable this feature. But everything has to work as before, if the option is not set.

**Comment by stevenvachon** [comment]
No breaking change is better than nothing, I guess, but it doesn't close this issue.

**Comment by nknapp** [comment]
You opened the issue as "*option* to disable {{{ }}}". If it's an option, it's not breaking. 

My main question is: do you expect the `<a href...>` in my fiddle to throw an error as well, if the option is set?

I would say " no ", because this is not something you would accidentally come up with. And I consider it essential to have this functionality. I would like to know what your use case is, that makes Handlebars.SafeString a security risk.

**Comment by stevenvachon** [comment]
Yes, an issue to disable `{{{ }}}` with the intent of avoiding accidental unescaped output. Using `{{ }}` with a helper (possibly third-party) that returns unescaped output (via `Handlebars.SafeString`) is even more accidental.

**Comment by nknapp** [comment]
First of all, let me repeat that I consider a compile-option to disable `{{{` useful, but I would not change the default behavior. If you want to make a PR, please make it for the `4.x` branch. I will merge everything into `master` later.

I would like to have additional opinions about disabling SafeString. Personally, I would expect that if someone knows how to use SafeString, he will also be aware that he has to call `Handlebars.Utils.escapeExpression` on the critical parts. This is also shown in the SafeString example on http://handlebarsjs.com/.

I would like to hear more opinions on this before adding it to the core.
